### PR TITLE
feat(EC22): Implementing Avoid Method Usage for Only Basic Operations

### PR DESCRIPTION
**Pull request by moabidi91**
### Description : <br>
Using methods for basic operations consumes additional system resources. 
The interpreter must in effect and solve the objects and then the methods, just to carry out these simple operations of the language.

### Implementation 

1. **What's a basic operation**  in Java?
 A "**basic operation**" involve the use of arithmetic, comparison, and logical operators.  <br>
These operators are used either in **Binary expression** or **Unary expression**  and also in **Conditional exprssion** using ternary operator.
<br>So **technically**, using Java Sonar's API, an expression which is instance of **BinaryExpressionTree** or **UnaryExpressionTree** or  is **Tree.Kind.CONDITIONAL_EXPRESSION** => represents a basic operation !

2. **What kind of methods should be targeted** ?
Each method having a return statement of a basic operation

3. **What this Rule do not cover ?**
The below example of code contains an assignment ecpression then a return statement  : 
```
public int add(int a, int b) {
  int c = a+b;
  return c;
}
```
As there is an **unnecessarily assignement** , it should be reported by another Rule.<br>
So respecting **Single Resp. Principle** of SOLID and regarding the scope of EC22 this practice wont be reported despite the presence of a basic operation.

Same case could be detected with **mutation**.

MicrosoftTeams-image (1)
MicrosoftTeams-image

### CODE REVIEW

**Comment by wokier** [comment]
Hi

I think this rule does not apply to Java, because the JVM does the optimisation at runtime.
Having a meaningful method for a basic operation is a recommendation for a good design for a rich domain.

**Comment by moabidi91** [comment]
@wokier Thanks for your review. <br>
- Could you please explain more how "JVM does the optimisation at runtime" for this use case ?

- I understand your point of view. Indeed, when it comes to simple operations that are not repeated frequently, it might not be worth creating a separate utility method for them, as it could decrease performance. So we need to find the right balance between duplicated code and appropriate use of utility methods (or readability). And so it's a sonar smell not a bug.

**Comment by MP-Aubay** [comment]
Issue #131

**Comment by wokier** [comment]
> * Could you please explain more how "JVM does the optimisation at runtime" for this use case ?
I know that the JVM is able to rearrange the program structure to simplify it.
In this case, it is called inlining. I have found this documentation that explains it https://blog.jdriven.com/2019/11/hotspot-jvm-optimizations/
I agree that this rule does apply to other language, especially interpreted ones, where the method will be an extra step.

> * I understand your point of view. Indeed, when it comes to simple operations that are not repeated frequently, it might not be worth creating a separate utility method for them, as it could decrease performance. So we need to find the right balance between duplicated code and appropriate use of utility methods (or readability). And so it's a sonar smell not a bug.
My conclusion is that this proposed rule would reduce the readability (in term of domain richness) and provide near to zero optimization.

**Comment by dedece35** [comment]
Hi @moabidi91,
please resolve conflict and check DoD list https://github.com/green-code-initiative/ecoCode-common/blob/main/doc/starter-pack.md#definition-of-done-of-a-pr

**Comment by dedece35** [comment]
Hi @moabidi91,
regarding previous discussions, I think we have to check if this rule is applicable for Java with tests based on decompilation to check real code built by compiler.

**Comment by dedece35** [comment]
I found another benchmark by baeldung ( a Java reference on web) : https://www.baeldung.com/jvm-method-inlining
It show us that optimization is done respecting some checks.

My local tests show us that by default there is no optimization.

Here is Java source : 
java

Here is decompiled Class file with IntelliJ : 
class-1

Here is decompiled Class file with another decompilation tool (JD-GUI app) : 
class-2

In conclusion, I don't think optimizations are systematic with JIT compiler and this optimization is maybe usefull.

**Comment by wokier** [comment]
Hello
Thanks for the researches
The baeldung page explains that the optimizations are done at runtime.
It can depends on the arguments given to the java command
You did'nt shown the same output as in the article.
Compile and decompile proves nothing.

**Comment by dedece35** [comment]
> Hello Thanks for the researches The baeldung page explains that the optimizations are done at runtime. It can depends on the arguments given to the java command You did'nt shown the same output as in the article. Compile and decompile proves nothing.

Hi @wokier,
sorry, my bad, you're rigth it's at runtime, I read too quickly the article ... I retry with options of article

**Comment by cyChop** [comment]
Many interesting facts and opinions. This is one of those times when finding the right balance between eco-friendliness and maintainability must be found.

I fully agree with both those statements:

> Having a meaningful method for a basic operation is a recommendation for a good design for a rich domain.

> So we need to find the right balance between duplicated code and appropriate use of utility methods (or readability). And so it's a sonar smell not a bug.

I'll also say this:

* When I look at developers using Sonar, I see three profiles:
  * Those who don't fix anything.
  * Those who have enough maturity to separate what's applicable from what's not.
  * Those who just do their best to fix everything, even smells and infos. I've been one of those and I've learned much in that time, but we need to be relevant especially for developers who will use this as a learning tool. We must help them be critical. That may be as simple as a rule explanation that invite to caution.

* Ultimately, having a method with a meaningful name stating the intention makes the code clearer. I'm not vouching for a sum method, but I'm sure we can find examples of code with a real functional meaning to the method that is a basic operation. \
For such cases, the method is important and must be preserved. If the intention/functional meaning is not obvious, the code will be less clear, so less maintainable, so it will become less green over time, as it evolves. I'm convinced a good code design is an important part of keeping the code green over time.

This is only food for thought, not a definite answer of course.

About testing which code is best:

> Compile and decompile proves nothing.

True, but a benchmark could be useful, if we can measure the CPU, memory and bandwidth used. It could help for all our rules, and if we can do something generic with Docker (for instance), we probably could apply it to all languages. It could also help us arbiter all rules, decide if they're relevant or not, provide figures to explain our rule choices and so on.

I've thought (months ago) of setting up something, probably with Docker and Scaphandre, to compare two snippets of code. ETSDiff might be a nice solution for that, too. Maybe something we could build on GreenFrame. Unfortunately, I've lacked the time until now, and things are not getting better on that front.

One final consideration we might have is: since which version this optimization has been implemented, and do we want to add a rule that applies only to the affected versions of Java? (I've wondered about this for loops and StringBuilders, since I've heard recently that the latest JVMs might have runtime optimizations for String concatenations in loops. I have my own answer for the String concatenation, but I'll keep it in mind.)

**Comment by dedece35** [comment]
decision : to measure and find use cases
